### PR TITLE
 mgr/nfs: allow dynamic update of cephfs nfs export

### PR DESCRIPTION
**Pull request by ajarr**
**mgr/nfs: allow dynamic update of cephfs nfs export**

mgr/nfs module's apply_export() method is used to update an existing
CephFS NFS export. The method always restarted the ganesha service (
ganesha server cluster) after updating the export object and notifying
the ganesha servers to reload their exports. The restart temporarily
affected the clients connections of all the exports served by the
ganesha servers.

It is not always necessary to restart the ganesha servers. Only
updating the export ID, path, or FSAL block of a CephFS NFS export
requires a restart. So modify apply_export() to only restart the
ganesha servers for such export updates.

The mgr/nfs module creates a FSAL ceph user with read-only or
read-write path restricted MDS caps for each export. To change the
access type of the CephFS NFS export, the MDS caps of the export's FSAL
ceph user must also be changed. Ganesha can dynamically enforce an
export's access type changes, but Ceph server daemons can't dynamically
enforce changes in caps of the Ceph clients. To allow dynamic updates
of CephFS NFS exports, always create a FSAL ceph users with read-write
path restricted MDS caps per export. Rely on the ganesha servers to
enforce the export access type changes for the NFS clients.

Fixes: https://tracker.ceph.com/issues/54025
Signed-off-by: Ramana Raja <rraja@redhat.com>

**mgr/nfs: Don't notify ganesha rados object**

... to update export if the ganesha server is going to be
restarted. The ganesha server restart will reload the exports.

Signed-off-by: Ramana Raja <rraja@redhat.com>

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by ajarr** [comment]
Need to add tests. Also need to test in multinode environment.

I created this draft to check whether this approach to allow dynamic updates of CephFS NFS exports is okay. Is it OK to always set the MDS caps of FSAL Ceph user to 'read-write', and rely on NFS-Ganesha to enforce the access type changes (ro/rw) for the NFS clients?

**Review by batrick — APPROVED** [review]
LGTM

**Comment by batrick** [comment]
draft -> needs-qa?

**Comment by gregsfortytwo** [comment]
retest this please

**Comment by vshankar** [comment]
retest this please

**Review by vshankar — CHANGES_REQUESTED** [review]
@ajarr 

```
nfs/module.py:8: note: In module imported here,
nfs/__init__.py:7: note: ... from here:
nfs/export.py: note: In member "_apply_export" of class "ExportMgr":
nfs/export.py:721: error: "FSAL" has no attribute "user_id"
```

**Comment by vshankar** [comment]
@ajarr ping?

**Comment by ajarr** [comment]
@ffilz would you mind reviewing the approach of the PR? I have explained the fixes in the commit messages. I especially would like your opinion on the following change mentioned in the commit message that may affect the security model of NFS-Ganesha gateways over CephFS to be used in OpenStack or kubernetes environments,
*"The mgr/nfs module creates a FSAL ceph user with read-only or read-write path restricted MDS caps for each export. To change the access type of the CephFS NFS export, the MDS caps of the export's FSAL ceph user must also be changed. Ganesha can dynamically enforce an export's access type changes, but Ceph server daemons can't dynamically enforce changes in caps of the Ceph clients. To allow dynamic updates of CephFS NFS exports, always create a FSAL ceph users with read-write path restricted MDS caps per export. Rely on the ganesha servers to enforce the export access type changes for the NFS clients."*

Use-case
Both OpenStack and OpenShift/kubernetes users would create CephFS subvolumes (CephFS subtrees) [1], export the CephFS subvolumes via NFS-Ganesha gateways by creating Ganesha exports using the ceph-mgr nfs module [2], and mount the exports in VMs or containers by connecting to the NFS Ganesha servers. The NFS-ganesha servers may be co-located with the Ceph MDS servers (common scenario in OpenStack deployments) and maybe in different subnets.


[1] https://docs.ceph.com/en/latest/cephfs/fs-volumes/#fs-volumes-and-subvolumes
[2] https://docs.ceph.com/en/latest/mgr/nfs/#create-cephfs-export

**Comment by ffilz** [comment]
Yes, that all looks good. If a Ganesha export is configured RO, Ganesha will fail any operations that attempt to modify data or metadata.

**Comment by ajarr** [comment]
@adk3798 @phlogistonjohn  it'd be great if some one from the orchestrators team reviews this PR. It's a useful PR for OpenStack manila and ODF.

**Comment by phlogistonjohn** [comment]
Thanks for alerting me to this PR.  I'll take a look soon!

File: src/pybind/mgr/nfs/export.py [hunk 1]
```diff
@@ -765,13 +757,33 @@ def _apply_export(
...
                     cluster_id,
                     new_export.path,
-                    new_export.access_type,
                     cast(str, new_fsal.fs_name),
                     cast(str, new_fsal.user_id)
                 )
                 new_fsal.cephx_key = old_fsal.cephx_key
             else:
+                expected_mds_caps = 'allow rw path={}'.format(new_export.path)
+                entity = new_export.fsal.user_id
```
[the remark this answers is not in the document]
Comment (ajarr) [reply src/pybind/mgr/nfs/export.py]: fixed! Thanks, @vshankar for spotting this!

**Comment by ajarr** [comment]
> Thanks for alerting me to this PR. I'll take a look soon!

Thanks, John!

**Review by vshankar — COMMENTED** [review]
@ajarr How hard is to have a test case for this change?

File: src/pybind/mgr/nfs/export.py [hunk 2]
```diff
@@ -518,19 +514,16 @@ def _update_user_id(
...
         osd_cap = 'allow rw pool={} namespace={}, allow rw tag cephfs data={}'.format(
             self.rados_pool, cluster_id, fs_name)
-        access_type = 'r' if access_type == 'RO' else 'rw'
 
         self.mgr.check_mon_command({
             'prefix': 'auth caps',
             'entity': f'client.{user_id}',
-            'caps': ['mon', 'allow r', 'osd', osd_cap, 'mds', 'allow {} path={}'.format(
-                access_type, path)],
+            'caps': ['mon', 'allow r', 'osd', osd_cap, 'mds', 'allow rw path={}'.format(path)],
```
[the remark this answers is not in the document]
Comment (ajarr) [reply src/pybind/mgr/nfs/export.py:532]: Done

File: src/pybind/mgr/nfs/export.py [hunk 3]
```diff
@@ -784,10 +796,13 @@ def _apply_export(
...
-        self._update_export(cluster_id, new_export)
 
-        # TODO: detect whether the update is such that a reload is sufficient
-        restart_nfs_service(self.mgr, new_export.cluster_id)
+        # TODO: detect whether the RGW export update is such that a reload is sufficient
+        if need_nfs_service_restart:
+            self._update_export(cluster_id, new_export, should_notify=False)
+            restart_nfs_service(self.mgr, new_export.cluster_id)
+        else:
+            self._update_export(cluster_id, new_export)
```
[the remark this answers is not in the document]
Comment (ajarr) [reply src/pybind/mgr/nfs/export.py]: Done

File: src/pybind/mgr/nfs/export.py [hunk 4]
```diff
@@ -353,11 +347,13 @@ def _fetch_export_obj(self, cluster_id: str, ex_id: int) -> Optional[Export]:
...
 
-    def _update_export(self, cluster_id: str, export: Export) -> None:
+    def _update_export(self, cluster_id: str, export: Export,
+                       should_notify: Optional[bool] = True) -> None:
         self.exports[cluster_id].append(export)
         self._rados(cluster_id).update_obj(
             GaneshaConfParser.write_block(export.to_export_block()),
-            export_obj_name(export.export_id), conf_obj_name(export.cluster_id))
+            export_obj_name(export.export_id), conf_obj_name(export.cluster_id),
+            should_notify=should_notify)
```
[the remark this answers is not in the document]
Comment (ajarr) [reply src/pybind/mgr/nfs/export.py]: I'm not sure what's wrong with this indentation.

**Comment by ajarr** [comment]
> @ajarr How hard is to have a test case for this change?

I may be able to add tests in qa/tasks/cephfs/test_nfs.py, which is run in qa/suites/orch/cephadm/workunits/task/test_nfs.yaml. I am not sure if cephadm workunit tests change be tested locally like cephfs workunits using vstartrunner. I may have to develop and test the qa changes in the teuthology machines.

File: src/pybind/mgr/nfs/export.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (vshankar) [reply src/pybind/mgr/nfs/export.py]: OK - it want't obvious. I was expecting `should_notify` aligned with `export.export_id`.

File: src/pybind/mgr/nfs/export.py [hunk 5]
```diff
@@ -353,11 +347,15 @@ def _fetch_export_obj(self, cluster_id: str, ex_id: int) -> Optional[Export]:
             log.exception("Export ID: %s not found", ex_id)
         return None
 
-    def _update_export(self, cluster_id: str, export: Export) -> None:
+    def _update_export(self, cluster_id: str, export: Export,
+                       need_nfs_service_restart: Optional[bool] = False) -> None:
```
Comment (vshankar) [inline src/pybind/mgr/nfs/export.py:351]: Does `need_nfs_service_restart` need to be an optional arg?

**Comment by ajarr** [comment]
```
export CEPH_QA_MAIL="rraja@redhat.com"; MACHINE_NAME=smithi; CEPH_BRANCH=wip-ajarr-dynamic-update-cephfs-nfs-export-mar13-2022
teuthology-suite -v --ceph-repo https://github.com/ceph/ceph-ci.git -c $CEPH_BRANCH -m $MACHINE_NAME -s orch:cephadm -p 50 -e $CEPH_QA_MAIL --suite-branch $CEPH_BRANCH --filter-all test_nfs
```
http://pulpito.front.sepia.ceph.com/rraja-2022-03-14_03:47:03-orch:cephadm-wip-ajarr-dynamic-update-cephfs-nfs-export-mar13-2022-distro-default-smithi/

**Comment by ajarr** [comment]
> @ajarr How hard is to have a test case for this change?

tests have been added to check nfs servers aren't restarted during dynamic update of CephFS NFS exports

**Comment by ajarr** [comment]
Other NFS related tests in orch:cephadm suite
```
$ teuthology-suite -v --ceph-repo https://github.com/ceph/ceph-ci.git --suite-repo https://github.com/ceph/ceph-ci.git  -c $CEPH_BRANCH -m $MACHINE_NAME -s orch:cephadm -p 50 -e $CEPH_QA_MAIL --suite-branch $CEPH_BRANCH --filter mgr-nfs-upgrade
```
http://pulpito.front.sepia.ceph.com/rraja-2022-03-14_03:52:10-orch:cephadm-wip-ajarr-dynamic-update-cephfs-nfs-export-mar13-2022-distro-default-smithi/

```
$ teuthology-suite -v --ceph-repo https://github.com/ceph/ceph-ci.git --suite-repo https://github.com/ceph/ceph-ci.git  -c $CEPH_BRANCH -m $MACHINE_NAME -s orch:cephadm -p 50 -e $CEPH_QA_MAIL --suite-branch $CEPH_BRANCH --filter-all smoke-roleless,centos_8.stream_container_tools_crun
```
http://pulpito.front.sepia.ceph.com/rraja-2022-03-14_03:54:12-orch:cephadm-wip-ajarr-dynamic-update-cephfs-nfs-export-mar13-2022-distro-default-smithi/

**Comment by ajarr** [comment]
@vshankar please take a look

**Comment by ajarr** [comment]
jenkins test make check

**Review by adk3798 — APPROVED** [review]
LGTM

File: qa/tasks/cephfs/test_nfs.py [hunk 6]
```diff
@@ -70,20 +70,41 @@ def _disable_nfs(self):
...
+        return self._orch_cmd(*args)
+
+    def _check_nfs_cluster_event(self, expected_event):
+        '''
+        Check whether an event occured during the lifetime of the NFS service
+        :param expected_event: event that was expected to occur
+        '''
+        # Wait few seconds for NFS daemons' status to be updated
+        wait_time = 10
+        while wait_time <= 120:
```
Comment (vshankar) [inline qa/tasks/cephfs/test_nfs.py:86]: While you are here, could this be changed to use `safe_while`?

**Review by vshankar — APPROVED** [review]
Minor nit. Otherwise looks good.

File: qa/tasks/cephfs/test_nfs.py (hunk 6 above)
Comment (ajarr) [reply qa/tasks/cephfs/test_nfs.py]: Done

**Comment by ajarr** [comment]
The latest was a minor update to test_nfs to use safe_while() when looping and check for nfs daemon event, https://github.com/ceph/ceph/pull/44795#discussion_r826568341

**Comment by vshankar** [comment]
jenkins test make check arm64

**Review by vshankar — COMMENTED** [review]
Looks good.

**Comment by ajarr** [comment]
```
$ export CEPH_QA_MAIL="rraja@redhat.com"; MACHINE_NAME=smithi; CEPH_BRANCH=wip-ajarr-dynamic-nfs-update-mar-15-2022
$ teuthology-suite -v --ceph-repo https://github.com/ceph/ceph-ci.git --suite-repo https://github.com/ceph/ceph-ci.git  -c $CEPH_BRANCH -m $MACHINE_NAME -s orch:cephadm -p 50 -e $CEPH_QA_MAIL --suite-branch $CEPH_BRANCH --filter-all test_nfs
```
http://pulpito.front.sepia.ceph.com/rraja-2022-03-16_15:48:07-orch:cephadm-wip-ajarr-dynamic-nfs-update-mar-15-2022-distro-default-smithi/


```
$ export CEPH_QA_MAIL="rraja@redhat.com"; MACHINE_NAME=smithi; CEPH_BRANCH=wip-ajarr-dynamic-nfs-update-mar-15-2022
$ teuthology-suite -v --ceph-repo https://github.com/ceph/ceph-ci.git --suite-repo https://github.com/ceph/ceph-ci.git  -c $CEPH_BRANCH -m $MACHINE_NAME -s orch:cephadm -p 50 -e $CEPH_QA_MAIL --suite-branch $CEPH_BRANCH --filter-all smoke-roleless,centos_8.stream_container_tools_crun
```
http://pulpito.front.sepia.ceph.com/rraja-2022-03-16_15:43:38-orch:cephadm-wip-ajarr-dynamic-nfs-update-mar-15-2022-distro-default-smithi/

```
$ export CEPH_QA_MAIL="rraja@redhat.com"; MACHINE_NAME=smithi; CEPH_BRANCH=wip-ajarr-dynamic-nfs-update-mar-15-2022
$ teuthology-suite -v --ceph-repo https://github.com/ceph/ceph-ci.git --suite-repo https://github.com/ceph/ceph-ci.git  -c $CEPH_BRANCH -m $MACHINE_NAME -s orch:cephadm -p 50 -e $CEPH_QA_MAIL --suite-branch $CEPH_BRANCH --filter mgr-nfs-upgrade
```
http://pulpito.front.sepia.ceph.com/rraja-2022-03-16_15:54:35-orch:cephadm-wip-ajarr-dynamic-nfs-update-mar-15-2022-distro-default-smithi/

**Comment by ajarr** [comment]
> 

@vshankar  All the above tests passed.

**Comment by vshankar** [comment]
jenkins test make check arm64

**Review by vshankar — APPROVED** [review]
Nice work!

**Pull request merged**